### PR TITLE
WPF Refactoring #1: Add abstraction layer for MEF

### PR DESCRIPTION
**Pull request by tom-englert**
### Problem
The WPF-UI implementation is very inconsistent, e.g. often favoring the UI-first anti-pattern, with parts of the logic implemented in the UI instead of the view model.

### Solution
I'd like to contribute iterative cleanup of some of the UI, starting with this PR that adds an abstraction layer to MEF, so DI can be used without a hard coded dependency to the container implementation. 
This will allow to use existing tooling and maybe later update the container to some more modern implementation (e.g .Microsoft.Extensions.DependencyInjection)

Please leave some feedback here if such a cleanup is appreciated.

### CODE REVIEW

**Comment by siegfriedpammer** [comment]
I appreciate your support.

Given that most parts of the UI are 13 years old and were hacked together on a weekend, I don't see a problem with the UI-first approach.

However, I agree that it's time to clean it up. I am really looking forward to having proper DI instead of static state everywhere.

@christophwille what do you think?

**Comment by christophwille** [comment]
We looked at cleaning up the non-decompiler parts (UI) on and off over the years, and only ever managed to do minor improvements. Help definitely welcome especially around standardizing approaches, modern best practices (MEF, DI).

**Comment by tom-englert** [comment]
OK, I'll do some review then and now, and try to provide step by step PR's, so it's trackable what has been changed.

- What about naming conventions, formatting and use of newer language features? I it OK to normalize/update this in the touched files?
- INotifyPropertyChanged implementation is duplicated everywhere, would using AOP (https://github.com/Fody/PropertyChanged) be an option?
- Should we create an issue or discussion as a central hook for the PRs?

**Comment by christophwille** [comment]
Fody is not an option.

All net80 targets can use latest language features, all I know is that primary ctors are a touchy subject (like please don't use those).

**Comment by siegfriedpammer** [comment]
> What about naming conventions, formatting and use of newer language features? I it OK to normalize/update this in the touched files?

* Naming convention is the usual .NET naming convention except that fields do not use any kind of prefix
* Use of newer language features: In general all language features can and should be used, however there are some exceptions:
   * primary constructors should only be used, if there is exactly one - the primary - constructor
   * file-scoped namespace declarations should not be applied to existing files - unless it's a refactoring that is applied to all files and the changes are all in one commit, which is then added to `.git-blame-ignore-revs`

**Comment by christophwille** [comment]
INPC - if base classes don't cut it, one of the current and active source generators? @siegfriedpammer or would that be a no-go?

**Comment by siegfriedpammer** [comment]
source generators are fine!

**Comment by christophwille** [comment]
> * Should we create an issue or discussion as a central hook for the PRs?

If you already have a plan, surely a good idea.

File: Directory.Packages.props
Comment (christophwille) [inline Directory.Packages.props:53]: Is that really a good idea? That will then be applied to all the VSIX projects, ilspycmd, the installer, et cetera. We have fewer MEF projects than non-MEF projects.

Comment (tom-englert) [reply Directory.Packages.props:53]: GlobalPackageReferences are compile-time only by default, and this is just an analyzer helping to make consistent export definitions.

**Comment by christophwille** [comment]
@tom-englert please let us know when you consider this "first refactoring" ready for review / merge

**Comment by tom-englert** [comment]
@christophwille this is a first consistent step and ready for review

File: fixCodeFormatting.bat
Comment (siegfriedpammer) [inline fixCodeFormatting.bat:1]: We already have https://github.com/icsharpcode/ILSpy/blob/master/BuildTools/format.bat for this purpose...

File: ICSharpCode.ILSpyX/Analyzers/Builtin/TypeExposedByAnalyzer.cs
Comment (siegfriedpammer) [inline ICSharpCode.ILSpyX/Analyzers/Builtin/TypeExposedByAnalyzer.cs:25]: This should be moved out of the namespace.

File: fixCodeFormatting.bat
Comment (tom-englert) [reply fixCodeFormatting.bat:1]: OK, with the pre-commit hook this works much better 😄

File: ICSharpCode.ILSpyX/Analyzers/Builtin/TypeExposedByAnalyzer.cs
Comment (tom-englert) [reply ICSharpCode.ILSpyX/Analyzers/Builtin/TypeExposedByAnalyzer.cs]: fixed

**Comment by siegfriedpammer** [comment]
May I ask you to clean your commits up before we merge this once you are done? For example 4a933f66a34bd7de597c03465ee2c6916361f105 and 22ad7958e7b38a2c73176b6c638260c2b227691d could be merged into one commit so that the file you added and then removed, never is visible in the commit history...

Thank you very much for helping us with the UI clean up! I really appreciate it!

**Comment by tom-englert** [comment]
@siegfriedpammer OK, I'll ping you when I completed this.

**Comment by tom-englert** [comment]
@siegfriedpammer History is now cleaned up

File: ICSharpCode.ILSpyX/Analyzers/Builtin/TypeExposedByAnalyzer.cs
Comment (siegfriedpammer) [inline ICSharpCode.ILSpyX/Analyzers/Builtin/TypeExposedByAnalyzer.cs:23]: Ah cool, because this class was moved to ILSpyX, which does not reference WPF, we no longer have to deal with the "Accessibility.dll/namespace ambiguity" 👍

File: ILSpy.sln.DotSettings
Comment (tom-englert) [reply ILSpy.sln.DotSettings:1]: Removing this would make the life harder for anyone using ReSharper, getting false positive warnings everywhere, making it impossible to write compliant code.
Is this what you want?

Comment (christophwille) [reply ILSpy.sln.DotSettings:1]: "make the life harder for anyone using X" where do we draw the line then on which tool is important enough for its custom configuration to go into the repo? Nobody in 10+ years ever asked for R#.

Comment (christophwille) [reply ILSpy.sln.DotSettings:1]: If you simply keep a copy for yourself and assume it unchanged?

Comment (tom-englert) [reply ILSpy.sln.DotSettings:1]: Added the missing rules to .editorconfig, so that's available to everybody now, independent of the tooling

Comment (ltrzesniewski) [reply ILSpy.sln.DotSettings:1]: > Nobody in 10+ years ever asked for R#.

That's because they all use Rider now 😅

On a more serious note, I'd like to vote for being as friendly as possible towards R#/Rider users. Almost everyone at my work prefers Rider over VS, and no one uses plain VS without R# anymore.

Comment (tom-englert) [reply ILSpy.sln.DotSettings:1]: @ltrzesniewski works now with .editorconfig instead of R#.settings, so that should be fine with Rider, too

Comment (christophwille) [reply ILSpy.sln.DotSettings:1]: > > Nobody in 10+ years ever asked for R#.
> 
> That's because they all use Rider now 😅
> 
> On a more serious note, I'd like to vote for being as friendly as possible towards R#/Rider users. Almost everyone at my work prefers Rider over VS, and no one uses plain VS without R# anymore.

and nobody on the core team uses anything besides VS...

**Review by siegfriedpammer — APPROVED** [review]

**Comment by siegfriedpammer** [comment]
Thank you very much for the great work!

**Pull request merged**
